File: backend/ai_tutor/agents/planner_agent.py
from __future__ import annotations
from typing import Optional, List, Dict, Any
from uuid import UUID
try:
    from supabase import Client
except Exception:  # pragma: no cover - optional dependency
    from typing import Any as Client
from agents.run_context import RunContextWrapper
from ai_tutor.core.llm import LLMClient

from ai_tutor.context import TutorContext
import os

# --- Get Supabase client dependency (needed for the tool) ---
from ai_tutor.dependencies import get_supabase_client

from functools import lru_cache, wraps
import asyncio
from ai_tutor.skills import skill
from ai_tutor.utils.tool_helpers import invoke # Import the invoke helper

from ai_tutor.core.schema import PlannerOutput

import logging
import traceback
import json
from ai_tutor.agents.models import FocusObjective

# Global cache for concept graph edges and last updated timestamp
_dag_cache = {
    "edges": None,
    "updated_at": None
}
_dag_cache_lock = asyncio.Lock()

async def _get_concept_graph_edges(supabase):
    """Fetch concept prerequisite edges from Supabase without relying on updated_at column."""
    response = supabase.table("concept_graph").select("prereq, concept").execute()
    if response and hasattr(response, 'data') and response.data:
        return response.data
    return []

logger = logging.getLogger(__name__)

# --- Define read_knowledge_base tool locally ---
@skill(cost="low") # This now creates a FunctionTool object named read_knowledge_base
async def read_knowledge_base(ctx: RunContextWrapper[TutorContext]) -> str:
    """Reads the Knowledge Base content stored in the Supabase 'folders' table associated with the current session's folder_id."""
    folder_id = ctx.context.folder_id
    user_id = ctx.context.user_id
    logger.info(f"Tool: read_knowledge_base called for folder {folder_id}")

    if not folder_id:
        logger.warning("Tool: read_knowledge_base - Folder ID not found in context.")
        return "Error: Folder ID not found in context."

    # --- ADD CHECK HERE ---
    # Check if analysis result with text is already in context from SessionManager loading
    if ctx.context.analysis_result and ctx.context.analysis_result.analysis_text:
        logger.info(f"Tool: read_knowledge_base - Found analysis text in context. Returning cached text.")
        return ctx.context.analysis_result.analysis_text

    try:
        logger.info(f"Tool: read_knowledge_base - Querying Supabase for folder {folder_id}")
        supabase = await get_supabase_client()
        response = supabase.table("folders").select("knowledge_base").eq("id", str(folder_id)).eq("user_id", user_id).maybe_single().execute()
        logger.info(f"Tool: read_knowledge_base - Supabase query completed for folder {folder_id}")
        if response.data and response.data.get("knowledge_base"):
            kb_content = response.data["knowledge_base"]
            logger.info(f"Tool: read_knowledge_base successful from Supabase. Content length: {len(kb_content)}")
            # Store it back into context in case it wasn't there (though SessionManager should handle this on load)
            if not ctx.context.analysis_result:
                 # Assuming AnalysisResult model exists and can be instantiated like this
                 from ai_tutor.agents.analyzer_agent import AnalysisResult
                 ctx.context.analysis_result = AnalysisResult(analysis_text=kb_content, vector_store_id=ctx.context.vector_store_id or "", key_concepts=[], key_terms={}, file_names=[])
            elif not ctx.context.analysis_result.analysis_text:
                 ctx.context.analysis_result.analysis_text = kb_content
            return kb_content
        else:
            logger.warning(f"Tool: read_knowledge_base - Knowledge Base not found for folder {folder_id} or query failed.")
            return f"Error: Knowledge Base not found for folder {folder_id}."
    except Exception as e:
        error_msg = f"Error reading Knowledge Base from Supabase for folder {folder_id}: {e}"
        logger.error(f"Tool: read_knowledge_base - {error_msg}\n{traceback.format_exc()}", exc_info=True)
        return error_msg

@skill(cost="low") # This now creates a FunctionTool object named dag_query
async def dag_query(ctx: RunContextWrapper[TutorContext], mastered: list[str]) -> list[str]:
    """Returns next learnable concepts based on the concept_graph table and user's mastered concepts."""
    logger.info(f"Tool: dag_query called. Mastered concepts: {mastered}")
    supabase = await get_supabase_client()
    # Fetch all prerequisite relationships between concepts, using cache
    edges = await _get_concept_graph_edges(supabase)
    logger.info(f"Tool: dag_query - Supabase query completed. Found {len(edges)} edges.")
    # Build prerequisite map: concept -> list of prereqs
    prereq_map: Dict[str, List[str]] = {}
    for e in edges:
        prereq_map.setdefault(e["concept"], []).append(e["prereq"])
    # Identify next learnable concepts: not yet mastered and all prereqs satisfied
    candidates = [c for c, prereqs in prereq_map.items() if c not in mastered and all(p in mastered for p in prereqs)]
    logger.info(f"Tool: dag_query - Calculated candidates: {candidates}")
    return candidates

async def determine_session_focus(ctx: TutorContext) -> FocusObjective:
    """Analyzes the KB and user state to determine the primary FocusObjective for the session."""
    logger.info(f"determine_session_focus started for session {ctx.session_id}")
    kb_text: Optional[str] = None
    next_concepts: Optional[List[str]] = None
    mastered: List[str] = []
    user_model_state_summary: str = "No user model state available."

    try:
        # 1. Retrieve knowledge base using invoke
        logger.info(f"determine_session_focus: Calling read_knowledge_base via invoke for session {ctx.session_id}")
        try:
            # --- FIX: Use invoke --- 
            # Pass the FunctionTool object and context
            kb_text = await invoke(read_knowledge_base, ctx=ctx)
            # --- END FIX ---
            logger.info(f"determine_session_focus: read_knowledge_base returned. Length: {len(kb_text) if kb_text else 'None'}")
            if kb_text and "Error:" in kb_text: # Check for errors returned by the tool
                 logger.error(f"determine_session_focus: Error from read_knowledge_base: {kb_text}")
                 raise ValueError(f"Failed to read knowledge base: {kb_text}")
        except Exception as tool_e:
            logger.error(f"determine_session_focus: Exception calling read_knowledge_base for session {ctx.session_id}: {tool_e}\n{traceback.format_exc()}", exc_info=True)
            raise # Re-raise to outer block

        # 2. Determine mastered concepts and summarize user model state
        try:
            if ctx.user_model_state and ctx.user_model_state.concepts:
                 mastered = [t for t, s in ctx.user_model_state.concepts.items() if s.mastery > 0.8 and s.confidence >= 5] # Assuming 0.8 mastery and 5 confidence means mastered
                 # Create a summary string for the prompt
                 state_items = []
                 for topic, state in ctx.user_model_state.concepts.items():
                     state_items.append(f"- {topic}: Mastery={state.mastery:.2f}, Confidence={state.confidence}, Attempts={state.attempts}")
                 if state_items:
                     user_model_state_summary = "Current user concept understanding:\n" + "\n".join(state_items)
                 else:
                     user_model_state_summary = "User has no tracked concepts yet."

            logger.info(f"determine_session_focus: Determined mastered concepts for session {ctx.session_id}: {mastered}")
            logger.info(f"determine_session_focus: User model state summary for prompt: {user_model_state_summary}")
        except Exception as mastery_e:
             logger.error(f"determine_session_focus: Error processing user model state for session {ctx.session_id}: {mastery_e}", exc_info=True)
             # Continue, maybe with empty mastered list and default summary

        # 3. Query DAG for next learnable concepts (using invoke)
        logger.info(f"determine_session_focus: Calling dag_query via invoke for session {ctx.session_id}")
        try:
            # --- FIX: Use invoke --- 
            # Pass the FunctionTool object, context, and mastered list as kwargs
            next_concepts = await invoke(dag_query, ctx=ctx, mastered=mastered)
            # --- END FIX ---
            logger.info(f"determine_session_focus: dag_query returned: {next_concepts}")
        except Exception as tool_e:
            logger.error(f"determine_session_focus: Exception calling dag_query for session {ctx.session_id}: {tool_e}\n{traceback.format_exc()}", exc_info=True)
            # Don't raise, planner can still function without DAG info
            next_concepts = None # Indicate that DAG info is unavailable

        # 4. Call LLM
        llm = LLMClient()
        system_msg = {
            "role": "system",
            "content": (
                "You are the Focus Planner agent. Your task is to analyze the provided Knowledge Base text (potentially truncated to show recent history), the user's current concept understanding (User Model State), and potential next concepts based on prerequisites (if available). "
                "Based on this analysis, select the single most important FocusObjective for the current tutoring session. Consider the importance of topics, prerequisites, and the user's progress. "
                "Output ONLY a single, valid JSON object conforming exactly to the FocusObjective Pydantic model schema. Ensure 'topic', 'learning_goal', 'priority' (integer 1-5), and 'target_mastery' fields are ALWAYS included. Do not add any commentary before or after the JSON object."
                "\n\nFocusObjective Schema:\n"
                "{\n"
                "  'topic': str,              // The primary topic or concept to focus on.\n"
                "  'learning_goal': str,      // A specific, measurable goal (e.g., 'Understand local vs global scope').\n"
                "  'priority': int,           // Priority 1-5 (5=highest). MANDATORY FIELD.\n"
                "  'relevant_concepts': List[str], // Optional list of related concepts from the KB.\n"
                "  'suggested_approach': Optional[str], // Optional hint (e.g., 'Needs examples').\n"
                "  'target_mastery': float,   // Target mastery level (e.g., 0.8). MANDATORY FIELD.\n"
                "  'initial_difficulty': Optional[str] // Optional initial difficulty (e.g., 'Medium').\n"
                "}"
            )
        }

        # --- KB Truncation Logic (Task E-2) --- #
        KB_INPUT_LIMIT_BYTES = 8000 # Approx 8kB ~ 2k tokens
        kb_prompt_content: str
        if kb_text and len(kb_text.encode('utf-8')) > KB_INPUT_LIMIT_BYTES:
             # Get the last N bytes (tail)
             kb_tail_bytes = kb_text.encode('utf-8')[-KB_INPUT_LIMIT_BYTES:]
             # Decode carefully, ignoring errors if split mid-character
             kb_snippet = kb_tail_bytes.decode('utf-8', errors='ignore')
             kb_prompt_content = f"... (Beginning of Knowledge Base truncated)\n\n{kb_snippet}"
             logger.warning(f"Knowledge base text for session {ctx.session_id} truncated to last ~{KB_INPUT_LIMIT_BYTES} bytes for planner prompt.")
        elif kb_text:
             kb_prompt_content = kb_text
        else:
             kb_prompt_content = "Knowledge Base is empty or unavailable."
        # --- End KB Truncation Logic ---

        dag_info = f"Suggested next learnable concepts based on prerequisites: {next_concepts}" if next_concepts is not None else "Prerequisite information (DAG) is not available for planning."

        messages = [
            system_msg,
            {"role": "user", "content": f"Knowledge Base Content (Recent History):\n{kb_prompt_content}"},
            {"role": "user", "content": user_model_state_summary},
            {"role": "user", "content": dag_info},
            {"role": "user", "content": "Select the single best FocusObjective for this session based on all available information. Respond only with the JSON object."}
        ]
        logger.info(f"determine_session_focus: Calling LLM for session {ctx.session_id}")
        try:
            response_text = await llm.chat(messages) # Removed response_format
            logger.info(f"determine_session_focus: LLM call completed for session {ctx.session_id}")
        except Exception as llm_e:
             logger.error(f"determine_session_focus: LLM call failed for session {ctx.session_id}: {llm_e}", exc_info=True)
             raise ValueError("LLM call failed during focus planning.") from llm_e

        # 5. Parse and Validate LLM Response
        try:
            logger.info(f"determine_session_focus: Attempting to parse LLM response JSON for {ctx.session_id}")
            # LLMClient with json_object mode should return parsed dict directly or raise error
            if isinstance(response_text, str): # Fallback if json_object mode didn't work as expected
                 # Clean up model response: strip markdown fences and extract JSON
                 import re
                 cleaned = response_text.strip()
                 if cleaned.startswith("```"):
                     cleaned = re.sub(r"^```[^\\n]*\\n", "", cleaned)
                 if cleaned.endswith("```"):
                     cleaned = re.sub(r"\\n```$", "", cleaned)
                 start = cleaned.find('{')
                 end = cleaned.rfind('}')
                 if start != -1 and end != -1:
                     json_str = cleaned[start:end+1]
                 else:
                    logger.error(f"determine_session_focus: Could not extract JSON object from LLM string response: {response_text}")
                    raise ValueError("LLM did not return a valid JSON object string.")
                 data = json.loads(json_str)
            elif isinstance(response_text, dict):
                data = response_text # Assume it's already parsed JSON
            else:
                 logger.error(f"determine_session_focus: Unexpected LLM response type: {type(response_text)}")
                 raise ValueError("Unexpected LLM response type during focus planning.")

            logger.info(f"determine_session_focus: JSON parsed/obtained successfully for {ctx.session_id}: {data}")

            # Validate and create FocusObjective model
            logger.info(f"determine_session_focus: Attempting to validate FocusObjective for {ctx.session_id}. Data: {data}")
            # Ensure target_mastery is present and a float before validation if needed
            if 'target_mastery' not in data or not isinstance(data.get('target_mastery'), (float, int)):
                 logger.warning(f"determine_session_focus: 'target_mastery' missing or not a number in LLM response for {ctx.session_id}. Setting default 0.8. Data: {data}")
                 data['target_mastery'] = 0.8 # Add default if missing, as it's required by spec

            # Ensure priority is present and an int before validation if needed
            if 'priority' not in data or not isinstance(data.get('priority'), int):
                 logger.warning(f"determine_session_focus: 'priority' missing or not an int in LLM response for {ctx.session_id}. Setting default 3. Data: {data}")
                 data['priority'] = 3 # Add default if missing

            focus_objective = FocusObjective.model_validate(data)
            logger.info(f"determine_session_focus: FocusObjective validated successfully for {ctx.session_id}")

        except json.JSONDecodeError as json_e:
            logger.error(f"determine_session_focus: Failed to parse LLM JSON response for session {ctx.session_id}: {json_e}. Response: {response_text}", exc_info=True)
            raise ValueError("Failed to parse planner output from LLM") from json_e
        except Exception as val_e: # Catch Pydantic validation errors etc.
             logger.error(f"determine_session_focus: Failed to validate FocusObjective for session {ctx.session_id}: {val_e}. Data: {data}", exc_info=True)
             raise ValueError("Failed to validate planner output") from val_e

        # 6. Store chosen objective in context
        ctx.current_focus_objective = focus_objective
        logger.info(f"determine_session_focus: Stored focus objective '{ctx.current_focus_objective.topic}' in context for session {ctx.session_id}")

        logger.info(f"determine_session_focus finished successfully for session {ctx.session_id}")
        return focus_objective

    except Exception as outer_e:
         # Catch any exception missed by inner blocks
         logger.critical(f"Unhandled exception within determine_session_focus for session {ctx.session_id}: {type(outer_e).__name__}: {outer_e}\n{traceback.format_exc()}", exc_info=True)
         # Consider returning a default/fallback FocusObjective or re-raise
         raise 

File: convex/agents/plannerAgent.ts
// Planner Agent - Session Focus Planning
// Ported from Python backend/ai_tutor/agents/planner_agent.py

import { BaseAgent, createAgentConfig, AgentUtils } from "./base";
import { AgentContext, AgentResponse, FocusObjective, PlannerOutput, ActionSpec } from "./types";
import { api } from "../_generated/api";
import type { ActionCtx } from "../_generated/server";

export interface PlannerInput {
  user_model_state?: any; // User's learning progress and mastery
  force_focus?: string; // Optional: force focus on specific topic
}

export interface UserModelState {
  concepts: Record<string, ConceptState>;
}

export interface ConceptState {
  mastery: number; // 0-1 scale
  confidence: number; // 1-10 scale
  attempts: number;
  last_interaction?: string;
}

export class PlannerAgent extends BaseAgent {
  private static readonly KB_INPUT_LIMIT_BYTES = 8000; // ~2k tokens
  private conceptGraphCache: { edges: any[] | null; updatedAt: number | null } = {
    edges: null,
    updatedAt: null
  };
  private convexCtx?: ActionCtx;

  constructor(apiKey?: string, convexCtx?: ActionCtx) {
    const config = createAgentConfig(
      "Focus Planner",
      "gpt-4-turbo-preview",
      {
        temperature: 0.5, // Balanced temperature for planning decisions
        max_tokens: 3000
      }
    );
    super(config, apiKey);
    this.convexCtx = convexCtx;
  }

  async execute(context: AgentContext, input: PlannerInput): Promise<AgentResponse<FocusObjective>> {
    AgentUtils.validateSessionContext(context);
    
    this.log("info", `Starting session focus planning for session: ${context.session_id}`);

    try {
      const { result, executionTime } = await this.measureExecution(async () => {
        return await this.determineSessionFocus(context, input);
      });

      this.log("info", `Session focus planning completed in ${executionTime}ms: ${result.topic}`);
      return this.createResponse(result, executionTime);

    } catch (error) {
      this.log("error", "Session focus planning failed", error);
      return this.createErrorResponse(`Session focus planning failed: ${error}`);
    }
  }

  private async determineSessionFocus(
    context: AgentContext,
    input: PlannerInput
  ): Promise<FocusObjective> {
    // 1. Read knowledge base
    const knowledgeBase = await this.readKnowledgeBase(context);
    
    // 2. Process user model state
    const { masteredConcepts, userModelSummary } = this.processUserModelState(input.user_model_state);
    
    // 3. Query concept graph for next learnable concepts
    const nextConcepts = await this.queryConceptGraph(masteredConcepts);
    
    // 4. Generate focus objective using LLM
    const focusObjective = await this.generateFocusObjective(
      context,
      knowledgeBase,
      userModelSummary,
      nextConcepts,
      input.force_focus
    );
    
    // 5. Store in context for future reference
    context.focus_objective = focusObjective;
    
    return focusObjective;
  }

  private async readKnowledgeBase(context: AgentContext): Promise<string> {
    try {
      // Check if analysis result is already in context
      if (context.analysis_result?.analysis_text) {
        this.log("info", "Found knowledge base in context, using cached version");
        return context.analysis_result.analysis_text;
      }

      // If folder_id is available, read from database
      if (context.folder_id && this.convexCtx) {
        this.log("info", `Reading knowledge base for folder: ${context.folder_id}`);
        
        try {
          const folderData = await this.convexCtx.runQuery(api.functions.getFolderEnhanced, {
            folderId: context.folder_id as any
          });
          
          if (folderData?.knowledge_base) {
            this.log("info", `Knowledge base found for folder: ${context.folder_id}`);
            return folderData.knowledge_base;
          }
        } catch (dbError) {
          this.log("error", "Failed to query folder knowledge base", dbError);
        }
      }

      this.log("warn", "No knowledge base available for planning");
      return "Knowledge base is empty or unavailable.";
      
    } catch (error) {
      this.log("error", "Failed to read knowledge base", error);
      return `Error reading knowledge base: ${error}`;
    }
  }

  private processUserModelState(userModelState?: UserModelState): {
    masteredConcepts: string[];
    userModelSummary: string;
  } {
    let masteredConcepts: string[] = [];
    let userModelSummary = "No user model state available.";

    try {
      if (userModelState?.concepts) {
        // Determine mastered concepts (mastery > 0.8 and confidence >= 5)
        masteredConcepts = Object.entries(userModelState.concepts)
          .filter(([_, state]) => state.mastery > 0.8 && state.confidence >= 5)
          .map(([topic, _]) => topic);

        // Create summary for LLM prompt
        const stateItems = Object.entries(userModelState.concepts).map(
          ([topic, state]) => 
            `- ${topic}: Mastery=${state.mastery.toFixed(2)}, Confidence=${state.confidence}, Attempts=${state.attempts}`
        );

        if (stateItems.length > 0) {
          userModelSummary = "Current user concept understanding:\n" + stateItems.join("\n");
        } else {
          userModelSummary = "User has no tracked concepts yet.";
        }
      }

      this.log("info", `Processed user model: ${masteredConcepts.length} mastered concepts`);
      
    } catch (error) {
      this.log("error", "Error processing user model state", error);
    }

    return { masteredConcepts, userModelSummary };
  }

  private async queryConceptGraph(masteredConcepts: string[]): Promise<string[] | null> {
    try {
      // Get concept graph edges (with caching)
      const edges = await this.getConceptGraphEdges();
      
      if (!edges || edges.length === 0) {
        this.log("warn", "No concept graph data available");
        return null;
      }

      // Build prerequisite map: concept -> list of prereqs
      const prereqMap: Record<string, string[]> = {};
      edges.forEach(edge => {
        if (!prereqMap[edge.concept]) {
          prereqMap[edge.concept] = [];
        }
        prereqMap[edge.concept].push(edge.prereq);
      });

      // Find next learnable concepts: not mastered and all prereqs satisfied
      const candidates = Object.entries(prereqMap)
        .filter(([concept, prereqs]) => 
          !masteredConcepts.includes(concept) && 
          prereqs.every(prereq => masteredConcepts.includes(prereq))
        )
        .map(([concept, _]) => concept);

      this.log("info", `Found ${candidates.length} next learnable concepts`);
      return candidates;
      
    } catch (error) {
      this.log("error", "Failed to query concept graph", error);
      return null;
    }
  }

  private async getConceptGraphEdges(): Promise<any[]> {
    try {
      // Simple caching mechanism
      if (this.conceptGraphCache.edges && 
          this.conceptGraphCache.updatedAt && 
          Date.now() - this.conceptGraphCache.updatedAt < 5 * 60 * 1000) { // 5 minutes cache
        this.log("info", "Using cached concept graph edges");
        return this.conceptGraphCache.edges;
      }

      // Query concept graph from Convex database
      if (this.convexCtx) {
        this.log("info", "Fetching concept graph edges from database");
        const edges = await this.convexCtx.runQuery(api.functions.getAllConceptGraphEdges, {});
        
        this.conceptGraphCache.edges = edges;
        this.conceptGraphCache.updatedAt = Date.now();
        
        this.log("info", `Fetched ${edges.length} concept graph edges from database`);
        return edges;
      } else {
        this.log("warn", "No Convex context available, using fallback mock data");
        // Fallback to mock data if no Convex context
        const edges = [
          { prereq: "basic_concepts", concept: "intermediate_concepts" },
          { prereq: "intermediate_concepts", concept: "advanced_concepts" }
        ];

        this.conceptGraphCache.edges = edges;
        this.conceptGraphCache.updatedAt = Date.now();
        
        return edges;
      }
      
    } catch (error) {
      this.log("error", "Failed to fetch concept graph edges", error);
      
      // Return empty array or mock data on error
      const fallbackEdges = [
        { prereq: "basic_concepts", concept: "intermediate_concepts" },
        { prereq: "intermediate_concepts", concept: "advanced_concepts" }
      ];
      
      return fallbackEdges;
    }
  }

  private async generateFocusObjective(
    context: AgentContext,
    knowledgeBase: string,
    userModelSummary: string,
    nextConcepts: string[] | null,
    forceFocus?: string
  ): Promise<FocusObjective> {
    // Truncate knowledge base if too large
    const kbContent = this.truncateKnowledgeBase(knowledgeBase);
    
    // Prepare DAG information
    const dagInfo = nextConcepts 
      ? `Suggested next learnable concepts based on prerequisites: ${nextConcepts.join(", ")}`
      : "Prerequisite information (DAG) is not available for planning.";

    // Add force focus information if provided
    const focusInstruction = forceFocus 
      ? `IMPORTANT: The user has specifically requested to focus on: ${forceFocus}. Prioritize this topic if it appears in the knowledge base or is pedagogically appropriate.`
      : "";

    const systemMessage = `You are the Focus Planner agent. Your task is to analyze the provided Knowledge Base text, the user's current concept understanding (User Model State), and potential next concepts based on prerequisites (if available). 

Based on this analysis, select the single most important FocusObjective for the current tutoring session. Consider the importance of topics, prerequisites, and the user's progress. 

${focusInstruction}

Output ONLY a single, valid JSON object conforming exactly to the FocusObjective schema. Ensure 'topic', 'learning_goal', 'priority' (integer 1-5), and 'target_mastery' fields are ALWAYS included. Do not add any commentary before or after the JSON object.

FocusObjective Schema:
{
  "topic": "string",              // The primary topic or concept to focus on.
  "learning_goal": "string",      // A specific, measurable goal (e.g., 'Understand local vs global scope').
  "priority": number,             // Priority 1-5 (5=highest). MANDATORY FIELD.
  "relevant_concepts": ["string"], // Optional list of related concepts from the KB.
  "suggested_approach": "string", // Optional hint (e.g., 'Needs examples').
  "target_mastery": number,       // Target mastery level (e.g., 0.8). MANDATORY FIELD.
  "initial_difficulty": "string"  // Optional initial difficulty (e.g., 'Medium').
}`;

    const response = await this.callOpenAI([
      {
        role: "system",
        content: systemMessage
      },
      {
        role: "user",
        content: `Knowledge Base Content (Recent History):\n${kbContent}`
      },
      {
        role: "user",
        content: userModelSummary
      },
      {
        role: "user",
        content: dagInfo
      },
      {
        role: "user",
        content: "Select the single best FocusObjective for this session based on all available information. Respond only with the JSON object."
      }
    ], {
      response_format: { type: "json_object" }
    });

    // Parse and validate the response
    const focusData = this.parseJSONResponse<any>(response.content);
    
    // Ensure required fields with defaults
    if (!focusData.target_mastery || typeof focusData.target_mastery !== "number") {
      this.log("warn", "target_mastery missing or invalid, setting default 0.8");
      focusData.target_mastery = 0.8;
    }

    if (!focusData.priority || typeof focusData.priority !== "number") {
      this.log("warn", "priority missing or invalid, setting default 3");
      focusData.priority = 3;
    }

    // Validate required fields
    this.validateRequiredFields(focusData, ["topic", "learning_goal", "priority", "target_mastery"]);

    // Create FocusObjective with proper defaults
    const focusObjective: FocusObjective = {
      topic: focusData.topic,
      learning_goal: focusData.learning_goal,
      priority: focusData.priority,
      relevant_concepts: focusData.relevant_concepts || [],
      suggested_approach: focusData.suggested_approach,
      target_mastery: focusData.target_mastery,
      initial_difficulty: focusData.initial_difficulty
    };

    this.log("info", `Generated focus objective: ${focusObjective.topic} (Priority: ${focusObjective.priority})`);
    return focusObjective;
  }

  private truncateKnowledgeBase(knowledgeBase: string): string {
    if (!knowledgeBase) {
      return "Knowledge Base is empty or unavailable.";
    }

    const textBytes = new TextEncoder().encode(knowledgeBase);
    if (textBytes.length <= PlannerAgent.KB_INPUT_LIMIT_BYTES) {
      return knowledgeBase;
    }

    // Get the last N bytes (tail)
    const truncatedBytes = textBytes.slice(-PlannerAgent.KB_INPUT_LIMIT_BYTES);
    const truncatedText = new TextDecoder().decode(truncatedBytes);
    
    this.log("warn", `Knowledge base truncated to last ~${PlannerAgent.KB_INPUT_LIMIT_BYTES} bytes`);
    return `... (Beginning of Knowledge Base truncated)\n\n${truncatedText}`;
  }
}

// Factory function for creating planner agent instances
export function createPlannerAgent(apiKey?: string, convexCtx?: ActionCtx): PlannerAgent {
  return new PlannerAgent(apiKey, convexCtx);
}

// Planning validation utilities
export class PlanningValidator {
  static validateFocusObjective(objective: FocusObjective): boolean {
    if (!objective.topic || typeof objective.topic !== "string") {
      throw new Error("FocusObjective must have a valid topic string");
    }
    
    if (!objective.learning_goal || typeof objective.learning_goal !== "string") {
      throw new Error("FocusObjective must have a valid learning_goal string");
    }

    if (typeof objective.priority !== "number" || objective.priority < 1 || objective.priority > 5) {
      throw new Error("FocusObjective priority must be a number between 1 and 5");
    }

    if (typeof objective.target_mastery !== "number" || objective.target_mastery < 0 || objective.target_mastery > 1) {
      throw new Error("FocusObjective target_mastery must be a number between 0 and 1");
    }

    return true;
  }

  static extractPlanningMetrics(objective: FocusObjective): {
    priority: number;
    targetMastery: number;
    conceptCount: number;
    hasApproach: boolean;
    hasDifficulty: boolean;
  } {
    return {
      priority: objective.priority,
      targetMastery: objective.target_mastery,
      conceptCount: objective.relevant_concepts?.length || 0,
      hasApproach: !!objective.suggested_approach,
      hasDifficulty: !!objective.initial_difficulty
    };
  }
}

// Helper function for session focus determination
export async function determineSessionFocus(
  context: AgentContext,
  userModelState?: UserModelState,
  forceFocus?: string,
  apiKey?: string
): Promise<FocusObjective | null> {
  try {
    const agent = createPlannerAgent(apiKey);
    const response = await agent.execute(context, { 
      user_model_state: userModelState,
      force_focus: forceFocus
    });
    
    if (response.success && response.data) {
      PlanningValidator.validateFocusObjective(response.data);
      return response.data;
    }
    
    console.error("Session focus planning failed:", response.error);
    return null;
    
  } catch (error) {
    console.error("Error in determineSessionFocus:", error);
    return null;
  }
} 